Variant of the section about coloring (#14)
adrhill/iclr2025-sparse-autodiff: _posts/2025-04-28-sparse-autodiff.md

</div>

The same idea can also be applied to reverse mode AD, as shown in Figure Y.

> Instead of leveraging orthogonal column, we exploit orthogonal rows.

Author: No problem. The main change here was to highlight that coloring isn't about `finding` orthogonal columns, but grouping them.
```suggestion
Instead of leveraging orthogonal columns, we rely on orthogonal rows.
```

### Sparsity pattern detection and coloring

Unfortunately, our initial assumption had a major flaw: 
since AD only gives us a composition of linear maps and linear maps are black-box functions,
the structure of the Jacobian is completely unknown.
In other words, **we cannot tell which rows and columns are orthogonal without first materializing a Jacobian matrix.**
But if we fully materialize a Jacobian via traditional AD, ASD isn't needed at all.

The solution to this problem is shown in Figure 10 (a):
in order to find orthogonal columns (or rows), we don't need to materialize the full Jacobian.
Instead, it is enough to **detect the sparsity pattern** of the Jacobian.
This binary-valued pattern contains enough information to deduce orthogonality.
From there, we use a **coloring algorithm** to group mutually orthogonal columns (or rows) together.

> Such a coloring can be visualized on Figure 10 (b), where the yellow columns will be evaluated together and the light blue ones too.

Author: ```suggestion
Such a coloring can be visualized on Figure 10 (b), where the yellow columns will be evaluated together (first JVP) and the light blue ones will be evaluated together (second JVP), for a total of 2 JVPs instead of 5.
```

    </div>
</div>
<div class="caption">
    Figure 10: The two elementary steps in ASD: (a) sparsity pattern detection, (b) coloring of the sparsity pattern.
    Figure 10: The first two steps of ASD: (a) sparsity pattern detection, (b) coloring of the sparsity pattern.
</div>

The solution to this problem is shown in Figure 10:
in order to find orthogonal columns (or rows), we don't need to materialize the full Jacobian.
Instead, it is enough to materialize a binary sparsity pattern of the Jacobian.
This pattern contains enough information to color it.
To sum up, ASD consists of four steps:

Performance is key: For one-off computations, these two steps need to be faster than the computation of columns or rows they allow us to skip. Otherwise, we didn't gain any performance.
As we will see in later benchmarks, this level of performance can be achieved.
Additionally, if we need to compute Jacobians multiple times and are able to reuse the sparsity pattern, 

> the cost of sparsity pattern detection and coloring can be amortized over time.

Author: Because these steps don't _need_ to be faster to be useful. Even if they are slower, sparse AD can still be beneficial due to pattern and coloring reuse. That's why I reformulated it slightly differently below.
I know that in our tracer paper we argue for single-use detection but here it's not the focus.

there are also many possible approaches to sparsity pattern detection,
each with their own advantages and tradeoffs.

The method we will present here corresponds to a binary forward-mode AD system 

> in which performance is gained by compressing matrix rows.

Author: I prefer not to use the term "compress" here because compression is already used to describe things related to coloring (especially as opposed to decompression).
If you find another word to describe this, feel free to add it again.

    Figure 13: Propagating an index set through a linear map to obtain a sparsity pattern.  
</div>

### Operator overloading

Instead of going into implementation details,
we want to provide some intuition on a typical implementation technique for sparsity detection: 

> **operator overloading**.

Reviewer: No, I'll explain this as alternative function evaluation (of which operator overloading would we one of many possible implementations).

Author: I reverted the change but I disagree with your chosen terminology. The term I have seen in the literature is "non-standard interpretation". I had never seen "alternative function evaluation" until today.

Reviewer: Sure, I’ll go with that. I should have clarified that pattern detection is still WIP and not yet ready for iteration. 👍

We will demonstrate this on a second toy example, the function

$$ f(\vx) = x_1 + x_2x_3 + \text{sgn}(x_4) .$$

> <!-- TODO: I don't see the graph here -->

Reviewer: Duplicate of L510

Author: ```suggestion
```

</div>

As discussed in the previous section,
all inputs are initialized with their respective singleton index sets.
The sparsity detection system must walk through the computational graph in topological order.
Each individual operator in the graph is "overloaded" to behave in a different way during sparsity detection.
Instead of computing the output value, 
it must correctly propagate and accumulate the index sets of its inputs, 
depending on whether it has a non-zero derivative or not.  

Figure 14 annotates index sets on the edges of the computational graph.

> Since addition and multiplication have non-zero derivatives with respect to both of their inputs, 

Reviewer: Please revert my description of the figure.
You don't even know what I'm talking about here since I haven't added the figure yet... 😅

Author: See [this comment](https://github.com/adrhill/iclr2025-sparse-autodiff/pull/14/files#r1845252142), my main issue is about terminology

the index sets of their inputs are accumulated and propagated. 
The sign function has a zero-valued derivatives for any input value. 
It therefore doesn't propagate the index set of its input. 
Instead, it returns an empty set.

> <!-- *TODO: switch to multivariate function, quickly discuss resulting Jacobian.* -->

Reviewer: Uncomment so this is visible on the draft website.

Author: ```suggestion
*TODO: switch to multivariate function, quickly discuss resulting Jacobian.*
```
There were some todos commented and some uncommented so I tried to unify but OK.

<!-- TODO -->

<!-- Coloring techniques? -->
## Coloring

Once we have detected a sparsity pattern, our next goal is to figure out how to group the columns (or rows) of the Jacobian.
The columns (or rows) in each group will be evaluated simultaneously with a single JVP (or VJP).
If they are mutually orthogonal, then this gives all the necessary information to retrieve every nonzero coefficient of the matrix.

### Graph formulation

Luckily, this can be reformulated as a graph coloring problem, which is very well studied.

> Let us build a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$ such that each column is a vertex of the graph, and two vertices are connected iff their respective columns share a non-zero index.

Reviewer: `\mathcal{V}` and `\mathcal{E}` should be introduced.

Author: ```suggestion
Let us build a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$ with vertex set $\mathcal{V}$ and edge set $\mathcal{E}$, such that each column is a vertex of the graph, and two vertices are connected iff their respective columns share a non-zero index.
```

> Put differently, an edge between vertices $j_1$ and $j_2$ means that columns $j_1$ and $j_2$ are not orthogonal.

Reviewer: What are $j_i$?

Author: two column indices

> There are more efficient representations, but we leave them aside.

Reviewer: Instead of "but we leave them aside", how about putting some references here?

Author: I will, but at the moment there are no references anywhere in the blog post. Let's put all of our content first and then I'll make a pass filling the blanks with citations.

We want to assign to each vertex $j$ a color $c(j)$, such that any two adjacent vertices $(j_1, j_2) \in \mathcal{E}$ have different colors $c(j_1) \neq c(j_2)$.
This constraint ensures that columns in the same color group are indeed orthogonal.

> If we can find a coloring which uses the smallest possible number of distinct colors, it will minimize the number of groups, and thus the computational cost of the AD step.

Reviewer: A figure might be nice, since this is still somewhat abstract.

Author: I suggested a visualization on Slack where we compare three different colorings of the matrix (a suboptimal one with 3 colors, an infeasible one with 2 colors and an optimal one with 2 colors). If we can draw the graph too, great, but even just the colored matrices would be good.

### Greedy algorithm

> Unfortunately, the graph coloring problem is NP-hard, meaning that there is (probably) no way to solve it polynomially for every instance.
> But there are efficient heuristics which generate good enough solutions in reasonable time.

Reviewer: ```suggestion
Unfortunately, the graph coloring problem is NP-hard, 
but there are efficient heuristics which generate good enough solutions in reasonable time.
```

Author: Not everyone knows what NP-hard means, I think this is important.
